File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-11.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        rows = len(matrix) 
        cols = len(matrix[0]) 

        dp = [ [ -1 for _ in range(cols) ] for _ in range(rows) ] 

        def dfs(r, c): 
            if dp[r][c] != -1: 
                return dp[r][c] 

            best = 1 

            for dr, dc in [(1, 0), (-1, 0), (0, 1), (0, -1)]: 
                nr, nc = r + dr, c + dc 

                if 0 <= nr < rows and 0 <= nc < cols and matrix[nr][nc] > matrix[r][c]: 
                    best = max(best, 1 + dfs(nr, nc)) 

            dp[r][c] = best 

            return best 

        for r in range(rows): 
            for c in range(cols): 
                dfs(r, c) 

        res = 0 

        for r in range(rows): 
            for c in range(cols): 
                res = max(res, dp[r][c]) 

        return res

File: Data Structures & Algorithms/search-for-word/submission-11.js
class Solution {
    /**
     * @param {character[][]} board
     * @param {string} word
     * @return {boolean}
     */ 

    exist(board, word) {
        let rows =  board.length 
        let cols = board[0].length 

        const dfs = (i, r, c, visited) => {
            if (i === word.length) return true 

            if (!( 
                r >= 0 && 
                c >= 0 && 
                r < rows && 
                c < cols && 
                board[r][c] === word[i] && 
                !visited.has(`${r}_${c}`)
            )) return false 

            visited.add(`${r}_${c}`) 

            const res = dfs(i + 1, r + 1, c, visited) || 
                        dfs(i + 1, r - 1, c, visited) || 
                        dfs(i + 1, r, c + 1, visited,) || 
                        dfs(i + 1, r, c - 1, visited) 

            visited.delete(`${r}_${c}`) 

            return res
        }  

        let visited = new Set()
        for (let r = 0; r < rows; r ++) {
            for (let c = 0; c < cols; c ++) {
                if (word[0] === board[r][c]) {
                    const res = dfs(0, r, c, visited) 
                    if (res) return true
                }
            }
        } 

        return false
    }
}
